### DISCUSSION TITLE
Is using next/head to set canonical urls good enough for SEO purposes?

### DISCUSSION THREAD

**Discussion by joakley-ftdi** [vercel/next.js#32341] [Help] [answered]
We have an e-commerce application that is using next/head to set the page-specific canonical urls.  In the page source, the canonical url has not been set, because next/head operates client-side after the page loads.  A consultant has warned us that not having the canonical url in the page source might negatively impact SEO because of the way bots behave.

Two questions:
(a) Is next/head a sufficient solution for bot traffic for SEO purposes, and if not
(b) is there a way to statically set the head in a page-specific way?

I looked into writing a custom document, but the custom document appears to be common to all pages and not possible to set a head value that's specific to a particular page.  Thanks in advance for your thoughts.

**Comment by joggienl** [comment]
a) if the values you put in the head are the correct values I would not see why it would "hurt" SEO.
b) you will have to use SSR or SSG/ISR to have it generated in the page. If you generate the page mostly in the browser it will not work. 

About canonical's specifically, if you are using dynamic pages (i.e. /blog/[slug].js) you cannot use `router.asPath` to get the canonical when using SSG or ISR. At build time it will put `/blog/[slug].js` in your canonical tag in stead of the slug value itself. There are issues where is being described why this happens (don't have them at hand now). When using SSR that should not matter I guess.

Take a look at [next-seo](https://www.npmjs.com/package/next-seo) which probably has a lot of helpful options for your use cases(s). There is also an example of how to use it in the nextjs repo: https://github.com/vercel/next.js/tree/canary/examples/with-next-seo

**Comment by joakley-ftdi** [comment, marked as answer]
There was a bug in our code that was causing the wrong value to be initialized in next/head.  We found it and have patched it.  Thank you.
